packages/webpack-dynamic-bundle-plugin/src/batch-dynamic-plugin.ts

```diff
@@ -0,0 +1,193 @@
+/*
+Copyright 2019 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import webpack from 'webpack';
+import {SyncWaterfallHook} from 'tapable';
+
+const {Template} = webpack;
+
+/**
+ * Webpack Plugin to batch dynamic chunk loading requests.
+ *
+ * <p>This plugin is compatible with both Webpack 4 and 5 APIs. It works by
+ * overriding parts of the webpack runtime. Instead of immediately generating
+ * script tags for each chunk, it uses a Promise to queue a task in the
+ * microtask queue that creates a single script tag for all chunks requested
+ * within the same synchronous code flow.
+ *
+ * Because of limitations in webpack's internal code structure, we always need
+ * to return some DOM node for each chunk. We return the actual script tag
+ * for the first call in a batch and comment nodes for subsequent calls.
+ *
+ * Most of the generated code in the Tempate is 1:1 the code generated by
+ * webpack's default implementation, divergence is clearly marked with DIFF.
+ */
+export default class DynamicBundlePlugin {
+  /**
+   * @param buildBatchEndpoint Helper to construct an asset URL from a serving
+   *                           prefix and a list of chunks ids.
+   */
+  constructor(private buildBatchEndpoint: (prefix: string, ids: number[]) => string) {
+  }
+
+  apply(compiler: webpack.Compiler) {
+    compiler.hooks.compilation.tap(DynamicBundlePlugin.name, (compilation) => {
+      let jsonpScriptHook = getWebpackJsonpHook(compilation);
+      jsonpScriptHook.tap(DynamicBundlePlugin.name, () => {
+        const {
+          crossOriginLoading,
+          chunkLoadTimeout,
+          jsonpScriptType
+        } = compilation.outputOptions;
+
+        const requireFn = '__webpack_require__';
+
+        const scriptNonce = `${requireFn}.nc`;
+        const activeBatch = `${requireFn}._ab`;
+        const publicPath = `${requireFn}.p`;
+
+        return Template.asString([
+          "var script;",
```

Comment:
Any reasons to not use ES6 syntax (const/let, arrow function, etc.) in the template?
Will the template be transpiled later or do we need to make it compatible for all browsers?

Seems like you at least use for...of... later so I assume we can use ES6 syntax?

Reply:
I think it will generally not be down-leveled since it doesn't count as an input source. Removing the `for...of` below. Nice catch!